### PR TITLE
ARROW-2758: [Plasma] Use Scope enum in Plasma

### PR DESCRIPTION
**Pull request by guoyuhong**
In this PR there are following changes:
1. add option "--scoped-enum" to Flat Buffer Compiler.
2. change the old-styled c++ enum to c++11 style. 

This change will unify the enum style in both Plasma and Ray. See Ray [Issue2121](https://github.com/ray-project/ray/issues/2121) and Ray [PR2194](https://github.com/ray-project/ray/pull/2194).

### CODE REVIEW

**Comment by wesm** [comment]
Looks like a good change

**Comment by guoyuhong** [comment]
I will reopen this PR after finishing the internal review.

**Comment by pcmoritz** [comment]
+1 LGTM

The test failure looks unrelated, I opened a JIRA for it here: https://issues.apache.org/jira/browse/ARROW-2757

**Review by pcmoritz — APPROVED** [review]

**Comment by pcmoritz** [comment]
@guoyuhong Can you create an account on the ARROW JIRA so I can assign this to you: https://issues.apache.org/jira/browse/ARROW-2758?

File: cpp/src/plasma/test/serialization_tests.cc [hunk 1]
```diff
@@ -353,13 +358,15 @@ TEST(PlasmaSerialization, WaitReply) {
   /* Create a map with two ObjectRequests in it. */
   ObjectRequestMap objects_in(num_objects_in);
   ObjectID id1 = ObjectID::from_random();
-  objects_in[id1] = ObjectRequest({id1, 0, ObjectStatus_Local});
```
Comment (guoyuhong) [inline cpp/src/plasma/test/serialization_tests.cc]: In this test case, the init value for the second argument is 0. However, in ObjectRequestType definition, there is no corresponding enum value to 0. I changed the value to ObjectRequestType::PLASMA_QUERY_LOCAL which equals to 1. The test has no problem. Will this change cause other problems? Maybe this is the benefit from scoped enum. We cannot assign arbitrary value to a enum type.

**Comment by guoyuhong** [comment]
Hi Philipp, I have create an account on the ARROW JIRA and left a comment is the page. I also left a comment in the test case change. Please take a look. Thanks.

**Pull request closed without merging**